### PR TITLE
Correcting handling of super-classes/interfaces on anonymous classes

### LINKED ISSUE #315: Anonymous classes don't seem to properly handle interfaces
```kotlin
interface Blah {
    fun method(): String
}

fun main(args: Array<String>) {
    println(TypeSpec.anonymousClassBuilder("")
            .addSuperinterface(Blah::class)
            .addFunction(FunSpec.builder("method").returns(String::class).addCode("return %S", "hello").build())
            .build().toString())

}
```

This prints:
```kotlin
object : Blah() {
  fun method(): kotlin.String = "hello"
}
```
Which does not compile because `Blah` does not have a constructor. The parenthesis should not be there. Looks like the problem is here : https://github.com/square/kotlinpoet/blob/master/src/main/java/com/squareup/kotlinpoet/TypeSpec.kt#L83

### PR DESCRIPTION
**Pull request by kiiadi**
Fixes #315

Also added helper method for constructing anonymous classes that have no arguments passed to a superclass  : `TypeSpect.anonymousClassBuilder()`

### CODE REVIEW

File: src/main/java/com/squareup/kotlinpoet/TypeSpec.kt [hunk 1]
```diff
@@ -79,10 +79,22 @@ class TypeSpec private constructor(builder: TypeSpec.Builder) {
...
+          listOf()
+        }
+
+        val allSuperTypes = supertype + superinterfaces.map { CodeBlock.of(" %T", it) }
+        if (allSuperTypes.size > 1) {
+          allSuperTypes.dropLast(1).forEach { codeWriter.emitCode(it).emit(",") }
+        }
+        if (allSuperTypes.isNotEmpty()) {
+          codeWriter.emitCode(allSuperTypes.last())
+        }
```
Comment (JakeWharton) [inline src/main/java/com/squareup/kotlinpoet/TypeSpec.kt]: These `if`s should be able to be replaced by the `joinToCode` helper.

Comment (kiiadi) [reply src/main/java/com/squareup/kotlinpoet/TypeSpec.kt]: 👍 thanks - I knew there must have been a better way!

Comment (kiiadi) [reply src/main/java/com/squareup/kotlinpoet/TypeSpec.kt]: Much neater - thanks!

File: src/main/java/com/squareup/kotlinpoet/TypeSpec.kt [hunk 2]
```diff
@@ -79,10 +79,18 @@ class TypeSpec private constructor(builder: TypeSpec.Builder) {
         }
         codeWriter.emit(" {\n")
       } else if (anonymousTypeArguments != null) {
-        val supertype = if (superinterfaces.isNotEmpty()) superinterfaces[0] else superclass
-        codeWriter.emitCode("object : %T(", supertype)
-        codeWriter.emitCode(anonymousTypeArguments)
-        codeWriter.emit(") {\n")
+
+        codeWriter.emitCode("object :")
+        val supertype = if (superclass != ANY) {
```
Comment (NightlyNexus) [inline src/main/java/com/squareup/kotlinpoet/TypeSpec.kt]: won't this break if i want
```kotlin
val FOO: Any = object : Any() {
}
```
for some reason?

how about `if (superclass != ANY || superinterfaces.isEmpty())`?

File: src/test/java/com/squareup/kotlinpoet/TypeSpecTest.kt [hunk 3]
```diff
@@ -154,6 +154,39 @@ class TypeSpecTest {
         |""".trimMargin())
   }
 
+  // https://github.com/square/kotlinpoet/issues/315
+  @Test fun anonymousClassWithMultipleSuperTypes() {
+    val superclass = ClassName("com.squareup.wire", "Message")
+    val anonymousClass = TypeSpec.anonymousClassBuilder()
+            .superclass(superclass)
```
Comment (Egorand) [inline src/test/java/com/squareup/kotlinpoet/TypeSpecTest.kt]: nit: continuation indent should be 4 spaces

File: src/main/java/com/squareup/kotlinpoet/TypeSpec.kt (hunk 2 above)
Comment (JakeWharton) [reply src/main/java/com/squareup/kotlinpoet/TypeSpec.kt]: That's equivalent to
```kotlin
val FOO: Any = object {
}
```
which is what will be emitted

Comment (NightlyNexus) [reply src/main/java/com/squareup/kotlinpoet/TypeSpec.kt]: oh, cool, I didn't know that was valid Kotlin. Thanks!

Comment (NightlyNexus) [reply src/main/java/com/squareup/kotlinpoet/TypeSpec.kt]: actually, won't this emit `...object : {...` (extra colon), or am i missing something again?

Comment (Egorand) [reply src/main/java/com/squareup/kotlinpoet/TypeSpec.kt]: Oh, right, `:` would be redundant

**Comment by kiiadi** [comment]
@Egorand have included a test case (and logic) to omit the extra `:` in the case that no super-types / super-interfaces are set.

**Comment by kiiadi** [comment]
Changed the logic so that the empty block emits as 
```kotlin
object { }
```
rather than 
```kotlin
object {
}
```

As this seems to better adhere to the style guides I've seen. Let me know if you agree.

Do I need to squash these down into one commit?

**Comment by Egorand** [comment]
@kiiadi agree with the style for brackets. Yes, please squash. Thanks!

**Comment by kiiadi** [comment]
Squashed.

**Comment by Egorand** [comment]
@JakeWharton can you take another look?

File: src/main/java/com/squareup/kotlinpoet/TypeSpec.kt [hunk 4]
```diff
@@ -79,10 +79,25 @@ class TypeSpec private constructor(builder: TypeSpec.Builder) {
...
+        }
+
+        val allSuperTypes = supertype + superinterfaces.map { CodeBlock.of(" %T", it) }
+
+        if (allSuperTypes.isNotEmpty()) {
+          codeWriter.emitCode(" :")
+          codeWriter.emitCode(allSuperTypes.joinToCode(","))
+        }
+        if (hasNoBody) {
+          codeWriter.emit(" { }")
```
Comment (swankjesse) [inline src/main/java/com/squareup/kotlinpoet/TypeSpec.kt]: is this to style? Typically in Java we always line break for simple consistency

File: src/test/java/com/squareup/kotlinpoet/TypeSpecTest.kt [hunk 5]
```diff
@@ -154,6 +154,58 @@ class TypeSpecTest {
...
+        ).build()
+
+    assertThat(toString(taco)).isEqualTo("""
+        |package com.squareup.tacos
+        |
+        |import com.squareup.wire.Message
+        |import java.lang.Runnable
+        |
+        |class Taco {
+        |    val NAME: Runnable = object : Message(), Runnable {
```
Comment (swankjesse) [inline src/test/java/com/squareup/kotlinpoet/TypeSpecTest.kt]: cool!

**Pull request merged**